nixos/tests/omnom.nix

```diff
@@ -26,17 +31,116 @@ in
+
+                  # log in
+                  driver.get(sys.argv[1])
+                  wait = WebDriverWait(driver, 3)
+                  wait.until(EC.url_changes(sys.argv[1]))
+
+                  # set up the addon
+                  driver.get(f"moz-extension://{uuid}/options.html")
+                  wait = WebDriverWait(driver, 3)
+                  wait.until(EC.visibility_of_element_located((By.CSS_SELECTOR, 'input[type="submit"]')))
+
+                  form = driver.find_element(By.CSS_SELECTOR, "form")
+                  token = form.find_element(By.CSS_SELECTOR, 'input[name="token"]')
+                  url = form.find_element(By.CSS_SELECTOR, 'input[name="url"]')
+                  submit = form.find_element(By.CSS_SELECTOR, 'input[type="submit"]')
+
+                  token.send_keys(sys.argv[2])
+                  url.send_keys(sys.argv[3])
+                  submit.click()
+
+                  # take snapshot
+                  # This will snapshot the popup page
+                  driver.get(f"moz-extension://{uuid}/popup.html")
+                  wait = WebDriverWait(driver, 3)
+                  wait.until(EC.visibility_of_element_located((By.CSS_SELECTOR, 'input[type="submit"]')))
+
+                  form = driver.find_element(By.CSS_SELECTOR, "form")
+                  submit = form.find_element(By.CSS_SELECTOR, 'input[type="submit"]')
+
+                  submit.click()
+
+                  driver.get(f"{sys.argv[3]}/my_bookmarks")
+                  wait = WebDriverWait(driver, 3)
+                  wait.until(EC.visibility_of_element_located((By.CSS_SELECTOR, "h3.title")))
+
+                  assert driver.find_element(By.CSS_SELECTOR, "h3.title").text == "My bookmarks (1)"
+
+                  driver.quit()
+                '';
+          in
+          [
+            pkgs.firefox-unwrapped
+            pkgs.geckodriver
+            seleniumScript
+            (pkgs.python3.withPackages (ps: with ps; [ selenium ]))
+          ];
       };
   };
 
-  # TODO: take a snapshot
   testScript =
     { nodes, ... }:
     # python
     ''
+      import re
+
+      url = "http://127.0.0.1:${toString servicePort}"
+
       server.start()
       server.wait_for_unit("omnom.service")
       server.wait_for_open_port(${toString servicePort})
-      server.succeed("curl -sf http://localhost:${toString servicePort}")
+      server.succeed(f"curl -sf {url}")
+
+      output = server.succeed("omnom create-user user user@example.com")
+      match = re.search(r"Visit (.+?) to sign in", output)
+      assert match is not None, "Login URL not found"
+      login_url = match[1].replace("0.0.0.0", "127.0.0.1")
+
+      output = server.succeed("omnom create-token user addon")
+      match = re.search(r"Token (.+?) created", output)
+      assert match is not None, "Addon token not found"
+      token = match[1]
+
+      server.succeed(f"selenium-script 'http://{login_url}' '{token}' '{url}'")
```

Comment:
This verifies that a snapshot happened, but we don't know how it looks like. Can we also copy the snapshot to the test output?